src/operator/bilinear_sampling.cu

```diff
@@ -0,0 +1,181 @@
+                                              const int i_w, const DType* grad,
+                                              const DType* data, const int o_n,
+                                              const int o_c, const int o_h,
+                                              const int o_w, DType* g_input,
+                                              const DType* grid_src,
+                                              DType* grad_grid) {
+  for (int index = (blockIdx.x + blockIdx.y * gridDim.x) * blockDim.x + threadIdx.x;
+       index < o_n * o_h * o_w;
+       index += blockDim.x * gridDim.x * gridDim.y) {
+    // (n, c, h, w) is the element in grad
+    int w = index % o_w;
+    int h = (index / o_w) % o_h;
+    int n = index / o_w / o_h;
+    DType top_left_y_gw = 0.0;
+    DType top_left_x_gw = 0.0;
+    index_t grid_src_index = n * o_h * o_w * 2 + h * o_w + w;
+    DType y_real = (*(grid_src + grid_src_index + o_h * o_w) + 1) * (i_h - 1) / 2;
+    DType x_real = (*(grid_src + grid_src_index) + 1) * (i_w - 1) / 2;
+    // the way in which cudnnSpatialTfSamplerBackward deals with boundaries
+    // is different from here
+    if (between(x_real, 0, i_w-1) && between(y_real, 0, i_h-1)) {
+      // within the boundaries
+      index_t top_left_y = static_cast<int>(floor(y_real));
+      index_t top_left_x = static_cast<int>(floor(x_real));
+      DType top_left_y_w = 1.0 - (y_real - top_left_y);
+      DType top_left_x_w = 1.0 - (x_real - top_left_x);
+      for (index_t c = 0; c < o_c; ++c) {
+        index_t grad_index = n * o_c * o_h * o_w + c * o_h * o_w + h * o_w + w;
+        index_t data_index = n * i_c * i_h * i_w + c * i_h * i_w + top_left_y * i_w + top_left_x;
+        // calc 4 vertex value in input data
+        DType top_left_v = *(data + data_index);
+        DType top_right_v = *(data + data_index + 1);
+        DType bottom_left_v = *(data + data_index + i_w);
+        DType bottom_right_v = *(data + data_index + i_w + 1);
+        // calc input grad
+        *(g_input + data_index) += *(grad + grad_index) * top_left_y_w * top_left_x_w;
+        *(g_input + data_index + 1) += *(grad + grad_index) * top_left_y_w * (1.0 - top_left_x_w);
+        *(g_input + data_index+ i_w) += *(grad + grad_index) * (1.0 - top_left_y_w) * top_left_x_w;
+        *(g_input + data_index+ i_w + 1) += *(grad + grad_index) * (1.0 - top_left_y_w) *
+                                            (1.0 - top_left_x_w);
+        // calc weight grad of top_left_w, then multiple -1 is the grad of grid_src
+        top_left_y_gw -= *(grad + grad_index) * (top_right_v - bottom_right_v +
+                         (top_left_v - top_right_v -
+                         bottom_left_v + bottom_right_v) * top_left_x_w);
+        top_left_x_gw -= *(grad + grad_index) * (bottom_left_v - bottom_right_v + (top_left_v -
+                         top_right_v - bottom_left_v + bottom_right_v) * top_left_y_w);

```

Comment:
I'm thinking if we need AtomicAdd to avoid possible computational error. I'm not very sure on this. @tornadomeet @piiswrong